### PR TITLE
Add enterprise AKS and workload security controls

### LINKED ISSUE #255: Implement enterprise AKS and container workload security rules
## Objective

Implement Azure Kubernetes Service and container workload security controls across Azure control-plane, Defender, and Kubernetes evidence sources.

## Scope

Implement these controls:

1. The AKS API server lacks approved IP restrictions.
2. An AKS cluster has no Kubernetes network policy.
3. Defender for Containers protection is disabled for an in-scope cluster.
4. AKS secrets lack required Key Vault or KMS-backed protection.
5. Secrets Store CSI secret rotation is disabled.
6. A workload permits privileged containers.
7. A workload permits unrestricted host network, host PID, or host IPC access.
8. A workload permits unrestricted hostPath volumes.
9. Kubernetes cluster-admin access is assigned too broadly.
10. A workload uses untrusted registries, mutable tags, or the latest image tag.

## Collection architecture

- Azure Resource Manager and AKS management APIs.
- Defender for Cloud plans and recommendations.
- Kubernetes API for namespaces, workloads, pod security context, RBAC, network policies, CSI configuration, and images.
- Key Vault/KMS integration metadata and cluster identity configuration.

## Required implementation behavior

- Handle private, stopped, unreachable, unsupported, and unauthorized clusters explicitly.
- Return NOT_APPLICABLE when no eligible cluster or workload exists.
- Return UNKNOWN for unavailable credentials, API errors, incomplete namespace discovery, or insufficient permissions.
- Return FAIL only from positive evidence of an unsafe setting.
- Include cluster, namespace, workload, container, image, subject, role, evidence, remediation, permissions, severity, confidence, and UNKNOWN reason.
- Do not treat an unavailable optional add-on as a failure without confirming scope.

## Delivery order

1. AKS API restrictions, network policy, Defender, and secret integration.
2. CSI rotation and pod security context rules.
3. HostPath, RBAC, registry trust, and image-tag rules.

## Acceptance criteria

- Every rule has collector, evaluator, remediation, documentation, and verified mappings.
- Tests cover secure and violating manifests, RBAC, network policies, CSI, Defender state, empty clusters, unreachable clusters, malformed data, and partial permissions.
- Unsupported or inaccessible clusters never produce false findings.
- No TBD framework mappings remain.
- Scanner, serialization, compliance, and CI-equivalent tests pass.
- An origin-branch PR references this issue.

### PR DESCRIPTION
**Pull request by ritiksah141**
## Summary

Implements the complete enterprise AKS and container workload security backlog from issue #255 with fifteen evidence-rich controls covering API exposure, Kubernetes network policy, Defender for Containers, secret protection, privileged and host access, cluster administration, and image supply chain policy.

## What changed

- Added `AZ-AKS-007` to require a private API server or organization-approved authorized IP ranges.
- Added `AZ-AKS-008` to require Azure, Calico, or Cilium network policy enforcement.
- Added `AZ-AKS-009` to detect eligible namespaces without a NetworkPolicy.
- Added `AZ-AKS-010` to require Microsoft Defender for Containers.
- Added `AZ-AKS-011` to detect native Kubernetes Secret references without Azure Key Vault KMS protection.
- Added `AZ-AKS-012` to require automatic rotation when the Key Vault Secrets Store CSI provider is enabled.
- Added `AZ-AKS-013` through `AZ-AKS-017` as separate controls for privileged containers, hostNetwork, hostPID, hostIPC, and hostPath volumes.
- Added `AZ-AKS-018` to detect cluster-admin bindings outside the approved subject allowlist.
- Added `AZ-AKS-019` through `AZ-AKS-021` as separate controls for untrusted registries, latest or implicit latest tags, and mutable non-digest image references.
- Added a cached AKS evidence collector for ARM, Defender for Cloud, Kubernetes namespaces, workloads, NetworkPolicies, and ClusterRoleBindings.
- Added strict policy loading through `OPENSHIELD_AKS_SECURITY_POLICY` and Kubernetes access through `OPENSHIELD_AKS_KUBECONFIG`.
- Added cluster, namespace, workload, container, image, subject, role, evidence, remediation, permissions, severity, confidence, and unknown reason metadata.
- Added an individual review-gated remediation playbook for every new rule.
- Added complete CIS, NIST CSF, ISO 27001, and SOC 2 mappings with explicit reviewed `N/A-*` CIS decisions where no direct benchmark mapping exists.
- Updated backend documentation, rule references, architecture guidance, contributor guidance, and the changelog without modifying `/website`.

## False-finding safeguards

- Empty AKS inventory is treated as not applicable.
- Missing policy, malformed policy, unavailable credentials, unauthorized Kubernetes access, and API failures are indeterminate and never create findings.
- Stopped clusters and unsupported provisioning states are reported as unknown rather than pass or fail.
- Failed namespace discovery is explicitly unknown, while namespace-specific failures are preserved as partial evidence.
- Partial collection retains valid positive evidence but never infers a missing NetworkPolicy for a namespace that could not be fully inspected.
- Private clusters satisfy the API exposure control without requiring public authorized IP ranges.
- Authorized IP ranges must be contained by an approved policy range, and malformed address evidence remains unknown.
- System and policy-excluded namespaces are omitted from workload and namespace policy findings.
- CSI rotation is not applicable when the CSI provider is disabled.
- Native Secret findings require observed workload references and do not infer secret use from cluster configuration alone.
- Controller-owned Pods are excluded to prevent duplicate findings already represented by their workload controller.
- Registry, cluster-admin, and namespace exceptions require explicit normalized policy entries.

## Verification

- Focused AKS collector, evaluator, Azure client, and policy suite: 94 passed.
- Full backend suite: 819 passed, 2 skipped.
- Total backend coverage: 86.85 percent, above the required 80 percent.
- Ruff check and formatting validation pass across the repository.
- Bandit reports no medium or high findings.
- Semgrep security-audit, OWASP Top Ten, Python, and JavaScript scans report no findings.
- Pip Audit reports no known vulnerabilities after the repository's existing tracked CI exclusions.
- Gitleaks reports no secrets in the staged change.
- Python package dependency validation reports no broken requirements.
- Compliance JSON and the fifteen-rule mapping matrix validate successfully.
- All sixteen AKS playbook scripts pass shell syntax validation.
- Git diff whitespace validation passes.
- Repository pre-commit checks pass.
- DCO verification passes for the commit.

## Type of change

- [x] New scan rules
- [x] Azure and Kubernetes evidence collectors
- [x] Remediation playbooks
- [x] Documentation
- [x] Compliance mappings
- [x] Regression and failure-path tests

## Testing

- [ ] Tested against a real Azure subscription and AKS cluster
- [x] Positive and secure evaluation paths covered
- [x] Empty inventory and not-applicable behavior covered
- [x] Stopped, unsupported, unauthorized, and unreachable states covered
- [x] Missing and malformed policy covered
- [x] Partial and incomplete namespace discovery covered
- [x] Local CI-equivalent backend and security checks pass
- [x] No hardcoded credentials or secrets

## Checklist

- [x] All fifteen issue #255 controls have collection, evaluation, evidence, remediation, mappings, and tests
- [x] Every rule has its own executable playbook
- [x] API and permission failures never become false PASS or FAIL results
- [x] Every commit includes a DCO `Signed-off-by` trailer
- [x] All four required compliance frameworks include `AZ-AKS-007` through `AZ-AKS-021`
- [x] No unresolved `TBD-*` framework mappings were introduced
- [x] Branch was created from the latest origin `dev`
- [x] Branch name follows the project convention

## Related issue

Closes #255

## Files to review

1. `scanner/aks_security.py` and `scanner/azure_client.py` - failure-aware ARM, Defender, and Kubernetes evidence collection.
2. `scanner/rules/_aks_enterprise_common.py` and `scanner/rules/az_aks_007.py` through `scanner/rules/az_aks_021.py` - strict policy evaluation and the fifteen independent controls.
3. `tests/test_aks_security_collector.py`, `tests/test_rules_aks_enterprise.py`, and `tests/test_azure_client_aks.py` - positive, secure, empty, malformed, partial, unsupported, and failure coverage.
4. `playbooks/cli/fix_az_aks_007.sh` through `playbooks/cli/fix_az_aks_021.sh` - individual review-gated remediation entry points.
5. `compliance/frameworks/*.json` - verified CIS, NIST CSF, ISO 27001, and SOC 2 mappings.
6. `config/aks-security-policy.example.json`, `docs/aks-security-rules.md`, `docs/rules-reference.md`, and `CHANGELOG.md` - policy schema, behavior, permissions, limitations, and release documentation.

### CODE REVIEW

**Review by TFT444 — CHANGES_REQUESTED** [review]
Good structure and solid failure-path coverage. Two blockers and two nits before merge.

**1. `policy_required` set too broad (blocker)**

`scanner/rules/_aks_enterprise_common.py:197`

Eight controls (`privileged`, `host_network`, `host_pid`, `host_ipc`, `host_path`, `secret_protection`, `network_policy_namespaces`, `latest_image`) return `[]` when `OPENSHIELD_AKS_SECURITY_POLICY` is not set, even though none of them need org-specific values. Each already handles `policy=None` safely downstream. Only four controls genuinely need a configured policy: `api_restrictions`, `cluster_admin`, `untrusted_registry`, and `mutable_image`. Remove the other eight from the set so workload rules run without a policy file.

**2. Docker Hub short-form images cause false-positive AZ-AKS-019 findings (blocker)**

`scanner/rules/_aks_enterprise_common.py:1581`

Images without a registry hostname (`nginx`, `redis:7`) fail the trusted-registry check even when `docker.io/` is in `trusted_registry_prefixes`, because `nginx.startswith(docker.io/)` is `False`. Normalize short-form names before the prefix match.

**3. Approved IP list rebuilt per cluster (efficiency)**

`scanner/rules/_aks_enterprise_common.py:1400`

`approved = [ipaddress.ip_network(item) for item in policy.approved_authorized_ip_ranges]` is inside the per-cluster loop. Hoist it above `for evidence in evidence_items`.

**4. Stale rule count in `docs/learn/index.html` (trivial)**

Heading says 95 Azure security rules but the paragraph body still reads 80 dynamic rules.

**Comment by ritiksah141** [comment]
Addressed the requested code changes and added regression coverage. I left the docs/learn/index.html rule count unchanged because the assignee handling docs/learn will update it.

**Comment by TFT444** [comment]
> Addressed the requested code changes and added regression coverage. I left the docs/learn/index.html rule count unchanged because the assignee handling docs/learn will update it.

cool make sence now its good to go for me approving it

**Review by TFT444 — DISMISSED** [review]

**Review by m-khan-97 — CHANGES_REQUESTED** [review]
Thanks for taking on this AKS security pack. There is a lot of good structure here, especially around preserving unknown and partial evidence. Before we merge it, I need us to close three gaps that could produce incorrect results in a real multi-cluster environment:

1. **Cluster evidence can be attached to the wrong AKS resource.** The collector selects kubeconfig contexts using only `cluster.name`. AKS names are scoped to a resource group, so two clusters in one subscription can share a name, and kubeconfig context names can also be customized. In that situation, both ARM resources may be evaluated through the same Kubernetes context. Please introduce an explicit, unambiguous resource-ID-to-context mapping, validate it, and treat missing or ambiguous mappings as UNKNOWN. Add a regression test with duplicate cluster names in different resource groups.

2. **The cluster-admin allowlist loses the ServiceAccount namespace.** ServiceAccounts are namespace-scoped, but AZ-AKS-018 compares only `Kind:name`. Allowing `ServiceAccount:builder` would therefore allow a `builder` account from every namespace. Please use a namespace-aware identity such as `ServiceAccount:<namespace>:<name>`, reject ambiguous policy entries, and test two same-named ServiceAccounts in different namespaces. Group and User identities can keep the appropriate non-namespaced form.

3. **Projected Secret volumes are not collected.** `_workload()` checks ordinary Secret volumes and environment references, but misses `volume.projected.sources[].secret.name`. A workload using only a projected Secret can therefore be treated as not applicable when KMS is disabled. Please collect those references and add a regression test.

The GitHub checks are green. My focused local run could not collect because this checkout is missing `azure.mgmt.authorization`; that environment limitation is separate from the production evidence cases above.

Please let me know once these cases are covered and I will re-review promptly.

**Comment by ritiksah141** [comment]
@m-khan-97, Addressed the tenant-scale review blockers in 9d6acb2. AKS contexts now use a validated resource-ID mapping, ServiceAccount allowlists are namespace-scoped, and projected Secret volumes are collected. Added regression coverage for each case. All local tests and vulnerability checks pass.

File: scanner/rules/_aks_enterprise_common.py [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+def scan_control(azure_client: Any, module: Mapping[str, Any], control: str) -> list[dict[str, Any]]:
+    rule_id = module["RULE_ID"]
+    evidence_items = azure_client.get_aks_security_posture()
+    if evidence_items is None:
+        logger.warning("%s: AKS inventory unavailable; result is UNKNOWN", rule_id)
+        return []
+    if not evidence_items:
+        logger.info("%s: no AKS clusters exist; result is NOT_APPLICABLE", rule_id)
+        return []
+    policy_required = control in {
```
Comment (parthrohit22) [inline scanner/rules/_aks_enterprise_common.py]: `policy_required` only covers api_restrictions/cluster_admin/untrusted_registry/mutable_image, so `policy` stays `None` for every other control - network_policy_namespaces, secret_protection, privileged, host_network/host_pid/host_ipc/host_path, and latest_image. That means `_workloads(evidence, policy)` at line ~189 (`if policy and namespace.lower() in policy.excluded_namespaces`) and the `network_policy_namespaces` branch's `excluded = policy.excluded_namespaces if policy else frozenset()` around line 387 can never apply the operator's configured namespace exclusions for those 8 rules, even though docs/aks-security-rules.md documents `excluded_namespaces` as a general policy field with no such caveat, and the PR description states "System and policy-excluded namespaces are omitted from workload and namespace policy findings" without qualification.

Suggest loading the policy for namespace-exclusion purposes independent of `policy_required` - e.g. always call `policy_from_env(rule_id)` and only gate the *stricter* "policy is mandatory to run this control at all" behavior behind `policy_required`, so namespace exclusion applies uniformly regardless of which control is running.

**Review by parthrohit22 — CHANGES_REQUESTED** [review]
Deep-reviewed _aks_enterprise_common.py (the shared evaluator all 15 new rules dispatch into), aks_security.py's collector, the azure_client.py diff, and traced logic through az_aks_007/009/011/013/018/019/020/021. This is a strong PR overall: the indeterminate-handling invariant is correctly applied across every control branch I walked (missing/None evidence, malformed IP ranges, malformed cluster-admin subjects, UNKNOWN status, missing/malformed policy all skip rather than assert compliance or a confirmed HIGH), the trusted-registry-prefix boundary bug this codebase apparently hit before (per the commit history) is correctly avoided here, RULE_IDs don't collide with anything existing, all 15 playbooks exist/pass bash -n and route through a shared confirmation-gated script that only auto-applies for the 4 rules with unambiguous low-blast-radius remediation, and framework mappings are complete and valid across all four compliance JSONs.

One real bug though: `excluded_namespaces` from the operator's AKS security policy is silently ignored by 8 of the 15 new rules, contradicting this PR's own documented safeguard. Verified with a standalone repro, not just by reading: built an AksClusterEvidence with `excluded_namespaces: ["istio-system"]` and a privileged workload with no NetworkPolicy in that namespace - az_aks_009.scan() and az_aks_013.scan() both still emit findings for it despite the explicit exclusion. Left an inline comment with the specific lines and root cause.

Why this got past the existing tests: every fixture that sets `excluded_namespaces` uses `["kube-system"]`, which is already unconditionally filtered at the collector level regardless of policy (`SYSTEM_NAMESPACES`) - so those tests pass whether or not the policy-based exclusion path actually runs. There's no test anywhere with a non-system excluded namespace for any of the 8 affected rules.

Concrete impact: an operator adds a service-mesh control-plane namespace (istio-system, linkerd) or a vendor-managed namespace to `excluded_namespaces`, per docs/aks-security-rules.md's own description of that field - findings keep firing on it anyway across network_policy_namespaces, secret_protection, privileged, host_network/pid/ipc/path, and latest_image. It's noise rather than a missed real risk (doesn't violate the "ambiguous evidence never confirmed HIGH" invariant), but it's broad and it's a documented claim that doesn't hold, which will train operators to distrust the tool's output on exactly the namespaces they explicitly told it to ignore.

Two non-blocking notes:
- `_normalize_image_registry` maps bare refs like `nginx:latest` to `docker.io/nginx:latest`, not Docker's canonical `docker.io/library/nginx:latest`. Internally consistent (tested), not a bug as long as operators write `docker.io/` in their policy - but a policy author who writes `docker.io/library/` per Docker's own convention would be surprised when official-image workloads keep getting flagged. Worth a one-line callout in the docs or the example policy comment.
- `_collect_kubernetes`'s six workload-kind calls per namespace (Deployment/StatefulSet/DaemonSet/Job/CronJob/Pod) share one try/except, so an early failure (e.g. RBAC denies listing DaemonSets) aborts the remaining calls for that namespace and marks it wholly partial, even if later calls would have succeeded. Conservative in the safe direction (under-reports, never a false HIGH), so not flagging as a bug - just noting it's untested at the collector level and slightly reduces detection completeness under partial RBAC.

Ran the full suite after installing the `kubernetes==36.0.0` dependency this PR adds (missing from this sandbox's .venv, an environment gap worth flagging in case CI needs it fresh too): 822 passed, 3 skipped, 0 failed. Test coverage for the new rules is substantive - compliant/noncompliant paths, empty inventory, inventory failure, unreachable cluster, malformed policy, missing policy split by policy-required-vs-independent rule sets, partial collection never inferring a missing NetworkPolicy, ServiceAccount-scoped cluster-admin allowlisting, IP-supernet containment, and registry prefix-boundary crossing are all exercised.

**Comment by m-khan-97** [comment]
@ritiksah141, Parth's namespace-exclusion finding is valid. Please update the shared evaluator so OPENSHIELD_AKS_SECURITY_POLICY is loaded when present for every control, while only the four policy-required controls fail closed when it is absent or invalid. That allows excluded_namespaces to apply consistently without making the policy-independent rules require configuration.\n\nPlease add regression coverage with a non-system namespace such as istio-system across the affected paths, including AZ-AKS-009 and at least the shared workload path used by AZ-AKS-011/013-017/020. Keep #293 draft while this is addressed and while #263/#308 remain dependencies.

**Review by TFT444 — DISMISSED** [review]
Re-reviewed. The explicit resource-ID-to-context mapping is in place: contexts are keyed by full resource ID (case-insensitive), ambiguous or missing mappings stay UNKNOWN, and a regression test for duplicate cluster names is present. My earlier concern is resolved. Approved.

**Comment by ritiksah141** [comment]
Dependencies #308 and #321 for #263 are merged into dev and included in this branch. Resolved the remaining namespace-exclusion blocker in 1afa662 by loading a supplied AKS policy for every control while keeping only four controls policy-required. Added non-system namespace regression coverage for AZ-AKS-009, AZ-AKS-011, AZ-AKS-013 through AZ-AKS-017, and AZ-AKS-020. Verification: 123 focused AKS tests passed; full backend suite passed with 1104 passed and 6 skipped; Ruff, formatting, and pre-commit checks passed. Marking ready for re-review.

**Comment by ritiksah141** [comment]
Follow-up: dev advanced with #336 after the earlier update. I merged it, preserved the new hash-locked dependency workflow, added Kubernetes to the runtime input, and made the Linux-only greenlet dependency explicit so runtime and development locks remain complete across platforms. All replacement hosted checks are green, including backend tests, lock validation, pip-audit, container build and Trivy, CodeQL, Semgrep, DCO, and CI Summary.

**Comment by ritiksah141** [comment]
@m-khan-97 @parthrohit22 @TFT444, PR #293 is ready for follow-up review. Dependencies #308 and #321 for #263 are merged and included, the namespace-exclusion blocker and latest dev conflicts are resolved, the review thread is closed, and all hosted checks including CI Summary, backend tests, DCO, CodeQL, Semgrep, pip-audit, dependency review, and Trivy are green. Please re-review when available.

**Review by parthrohit22 — CHANGES_REQUESTED** [review]
Thanks for the substantial AKS security coverage in this PR, and for addressing the earlier namespace-exclusion feedback. I’m requesting changes before merge to ensure the remediation guidance and documentation match the scanner’s real behavior.

In particular, the AZ-AKS-011 playbook currently enables the Key Vault CSI add-on but does not resolve existing native Kubernetes Secret usage; the guidance should treat this as a prerequisite and include the required workload migration steps. Please also clarify that only the policy-dependent controls become `UNKNOWN` when policy is unavailable, while the remaining controls can still report evidence-based findings.

**Comment by ritiksah141** [comment]
@parthrohit22, resolved in `2fdd190`.

The AZ-AKS-011 playbook now makes clear that enabling the Key Vault Secrets Store CSI add-on is only a prerequisite, not a complete remediation. It provides the required workload migration steps: configure Workload Identity and least-privilege Key Vault access, create and test a `SecretProviderClass`, mount the CSI volume, remove native Secret volume/projected Secret/`env`/`envFrom` references, restart and validate the workload, then re-run OpenShield. The guidance also preserves Key Vault KMS as the alternative when native Kubernetes Secrets must remain.

The AKS documentation now states that only AZ-AKS-007, AZ-AKS-018, AZ-AKS-019, and AZ-AKS-021 become UNKNOWN when policy is unavailable; the remaining controls continue evaluating available ARM and Kubernetes evidence. Regression tests cover both corrections.

Local verification: 125 focused AKS tests passed; full suite 1,118 passed and 6 skipped; Ruff, formatting, shell syntax, pre-commit checks, and DCO passed. Please re-review when ready.

**Comment by Vishnu2707** [comment]
@parthrohit22 @m-khan-97 - Please review this PR, if the blockers are resolved, kindly approve and merge.

**Review by parthrohit22 — APPROVED** [review]
Re-reviewed the remediation and documentation changes. The AZ-AKS-011 playbook now makes the required workload migration explicit and preserves Key Vault KMS as the alternative. The documentation now accurately states that only AZ-AKS-007, AZ-AKS-018, AZ-AKS-019, and AZ-AKS-021 are UNKNOWN when policy is unavailable; policy-independent controls continue evaluating available evidence. The branch is current with dev, mergeable, and hosted checks are green. Approved.

**Review by TFT444 — APPROVED** [review]
Re-reviewing against the two outstanding concerns from @m-khan-97:

1. **ServiceAccount namespace scope (AZ-AKS-018 allowlist):** The diff explicitly documents and implements `ServiceAccount:namespace:name` format — confirmed at the allowlist documentation line. Concern addressed.
2. **Projected Secret volumes in `_workload()`:** The collector now iterates `volume.projected.sources[]` and extracts `secret.name` for each source entry. Concern addressed.

Both earlier concerns from @m-khan-97 are resolved in code. All other blockers from prior rounds were fixed and verified. CI is green (822 tests, 86.85% coverage). Approving.

**Comment by ritiksah141** [comment]
@m-khan-97, could you please re-review this PR as other's have approved and only pending by you. So, if its okay then please approve or lemme know if more changes needed

**Review by m-khan-97 — APPROVED** [review]
I re-reviewed the updated head after bringing it onto current dev. My original three blockers are resolved: AKS resources use validated resource-ID-to-context mapping with ambiguous or missing mappings remaining UNKNOWN; cluster-admin ServiceAccount allowlists are namespace-scoped; and projected Secret volumes are collected. I also verified the later namespace-exclusion and remediation/documentation corrections remain intact after the merge.

The focused local run passed 121 tests; four collector tests could not import because this reusable review environment lacks the kubernetes package. The dependency-complete hosted backend suite passed on this exact head, along with CI Summary, rule validation, DCO, CodeQL, Semgrep, pip-audit, dependency review, Trivy, and every other required check. Approved.
